### ISSUE TITLE
`network.common.scalar` → `tenet.full_trace`: the qdim-weighted categorical trace and its sibling `inner` promoted into `tenet.ops`, with the name and the rank-0 question decided once

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/TeNeT-py#126]
**Goal**

Three issues have now deferred the same thing by name. #114 refused it inside a move-PR ("the bigger step — a qdim-weighted trace in `tenet.ops` next to `trace` — is a public-API design question and gets its own issue"). #120 reviewed all 24 `tenet.network` names, called `scalar` "the weakest keep here — `scalar(t)` on a rank-2 map silently *traces*, which the name does not say", and parked the `scalar` → `qdim_trace` rename *to* this issue so the name is decided once. #122 shipped `expectation_1site`/`expectation_2site`, noted they made it "the fourth and fifth caller of a qdim-weighted trace that has no public spelling — so the pressure is real and growing", and deferred again.

This is that issue. It settles four things and ships them in one PR: the **name**, the **semantics** (what shape it accepts and what it returns), whether **`inner`** rides along, and how the design stance "scalars are spelled with boundary legs" is squared with a function that returns a bare backend scalar.

It is a promotion of arithmetic that already exists and already runs under `grad`. No number moves.

**The gap, verified in this repo rather than argued**

Four facts, each produced by running the code on `main` today:

- `network.common.scalar(t)` on a rank-2 SU(2) map equals `np.trace(t.to_dense())` to the last bit. The qdim weight is what makes that identity hold, exactly as it is for `norm`.
- `tenet.trace(t, (0, 1))` on that same rank-2 tensor **raises**: `"tensordot: this contraction leaves no free leg, and TensorStructure needs at least one leg (a rank-0 SymmetricTensor does not exist)"`. The full trace of an endomorphism is not reachable through the public API at all — that is the hole five call sites are working around.
- `scalar` is **already rank-general**, despite a docstring that says "rank-2 map". On a rank-4 `(V OUT, V OUT | V IN, V IN)` tensor it returns exactly `np.einsum("abab->", t.to_dense())` — the trace of the map view, codomain against domain in order. `to_matrices` does the work and does not care about rank.
- `scalar` is **silently wrong on a non-square map**. On `(V OUT, W IN)` with `V != W` it returns a number (`-1.2986…`); on a same-side `(V OUT, V OUT)` pair it returns a number (`1.0531…`). Both are `np.trace` summing the leading diagonal of a rectangular matrix. No caller does this today, but nothing stops one, and promoting the function to public API without a refusal would bless it.

That last one is the strongest single argument for the promotion. The fix — `_check_square`, which `eigh`, `expm`, `eig` and `eigvals` already share — is one call, and it is not available where the function currently lives.

**The bar, taken from #120 unchanged**

A name must clear one of two bars: *it would confuse a newcomer arriving from tenpy/YASTN/quimb*, or *it is actively misleading about what the thing does*. Difference from the references is not a bar. Aliases are forbidden — one name per thing, and the old spelling is deleted in the same PR, not soft-landed.

**Decision 1 — the name is `tenet.full_trace`**

| candidate | verdict |
|---|---|
| `scalar` (today) | **refused.** #120's own finding stands: it silently traces and the name does not say so. It also collides conceptually with `ops/basic.py::_scalar`, the *scalar-operand* checker, and with `examples/vmc_mps.py::scalar`, which is a different function. |
| `qdim_trace` (#120's parked candidate) | **refused, and this is the reversal this issue owes an argument.** `tenet.norm` is already qdim-weighted and is not called `qdim_norm`; `tenet.trace`, the partial one, is qdim-correct too. The weight is not a variant of the operation, it *is* the operation — so a `qdim_` prefix implies an unweighted sibling that does not and must not exist, and it names the one thing the two traces have in common instead of the thing that differs. It flunks bar 2 for the same reason `scalar` does. |
| `tenet.trace(t, axes=None)` overload | **refused, on the repo's own recorded precedent.** `ops/contraction.py`:274-276 already says it, in the error text a caller sees: *"Returning a bare backend scalar would make the return type depend on the arguments — tenet.norm is the precedent for scalars leaving the tensor world explicitly."* A keyword that flips tensor-vs-scalar is the #64-shaped smell, and it is refused here by the function it would be bolted onto. |
| `tr` (TensorKit's spelling for the categorical trace of an endomorphism) | **refused.** `tenet.tr` next to `tenet.trace` is `canonize`/`canonicalize` again — two near-homographic names for related-but-different operations, the exact failure #120 declined to pay for. The concept is right and the survey should record it; the two-letter spelling is a C++/Julia-namespace convenience Python does not need. |
| `trace_all`, `trace_full`, `close` | **refused.** `all`/`full` postfix reads as an option on `trace`; `close` names the diagram, not the value. |
| **`full_trace`** | **adopted.** It names the actual distinction — *closed* versus *open*, scalar versus tensor — rather than the weight both share. It survives someone arriving from any of the three references without a doc lookup, and it sorts next to nothing it can be confused with. |

**Decision 2 — the design principle, stated so it stops being re-litigated**

`docs/design.md` "Scalars are spelled with boundary legs" is about **networks**: a contraction that leaves no free leg is refused, boundary legs are trivial-sector legs of dimension one, and the rank-0 tensor stays a deliberately unanswered question. `tenet.norm` has always coexisted with that, and it is the precedent to generalize. Written once, in `docs/design.md` and in the new docstring:

> **Open diagrams are tensors; closed diagrams exit to backend scalars, explicitly and by name.** `tensordot`, `einsum` and `trace` never return a scalar — a contraction that closes a network is a `ValueError`, and a `SymmetricTensor` still has no rank 0. Leaving the tensor world is a separate, named call — `norm`, `full_trace`, `inner` — which returns the backend's own scalar and is therefore traceable and differentiable.

`full_trace` **does not** give `SymmetricTensor` the rank-0 exit it lacks, which is the question #122 flagged as blocking. It gives the *user* an exit, at a name, with the categorical weight already applied — the same thing `norm` has done since Milestone 2. The rank-0 tensor question is untouched and stays open.

**Decision 3 — the signature**

```python
def full_trace(t: SymmetricTensor) -> Any:
    """``sum_c qdim(c) · tr(M_c)`` — the categorical trace of an endomorphism, a backend scalar."""
```

- **Any rank with a square map**, not rank-2 only. That is what the code already does (verified above) and what `expectation_2site`'s successors will want; restricting to rank 2 would be inventing a limitation to document.
- **The pairing is the map view**: codomain against domain, in order — `as_map(t)`, the same view `eigh`, `expm`, `svd` and `to_matrices` act through. Not an axes argument. This is why the rank-4 answer is `"abab->"` and not `"abba->"`; `tenet.trace(t, axes)` remains the way to close one *chosen* pair.
- **`_check_square` on entry**, the identical refusal `eigh`/`expm`/`eig`/`eigvals` share, with `"full_trace"` as the caller name. This is the one behavioural change in the PR and it converts today's silent wrong answer into a message that names the offending axis.
- **`requires(provider, QuantumDimension)`**, for parity with `norm`. Today a provider without `qdim` gets an `AttributeError` from a private attribute read; after promotion it gets `CapabilityError`.
- **Returns the backend's own scalar. No `float()`.** `norm`'s comment is the specification: concretizing makes the function unusable under `jit`/`grad`/`vmap`, and `examples/ctmrg.py::log_kappa` differentiates through this today. Callers wanting a Python float keep writing `float(tenet.full_trace(t))`, which every current caller already does.
- **A block-free tensor returns `0.0`**, matching `norm`'s explicit empty case rather than `sum()`'s bare `0`.

**Decision 4 — `inner` is promoted too**

Not as a convenience, and not because it is `full_trace`'s sibling. It is **`norm`'s** sibling, and it is currently a pure tensor operation living in the driver layer:

- `inner(a, b)` contains no MPS, no DMRG and no network concept whatsoever — it is `adjoint`, one `einsum`, one trace. It is in `tenet.network` for the historical reason that `lanczos` was the first thing to need it, which is a layering accident, not a design.
- `norm(a)**2 == inner(a, a)` is already an enforced identity (`tests/integration/test_dmrg.py`:193-195). Half of a pair that a test binds together being public while the other half sits two packages down is the inconsistency worth removing.
- The user-facing consequence today: anyone needing `<a|b>` for two symmetric tensors must import from `tenet.network`, i.e. reach into the DMRG driver layer for a vector-space primitive.
- Cost: two lines moved, one new `__all__` entry, one caller re-pointed.

Its one documented limit rides along in the docstring: the `string.ascii_lowercase` labelling caps it at rank 26, above which `einsum`'s own parser raises with a clear message. No fix — a rank-27 tensor has other problems.

**What lands, by file**

```
src/tenet/ops/contraction.py
    full_trace(t)                     # moved body from network/common.py::scalar,
                                      #   + _check_square + requires(QuantumDimension)
    inner(a, b)                       # moved body from network/common.py::inner
src/tenet/map_view.py
    check_square(m, caller)           # moved from ops/linalg.py::_check_square, verbatim
src/tenet/ops/linalg.py
    imports check_square from map_view; eigh/expm/eig/eigvals otherwise untouched
src/tenet/ops/__init__.py, src/tenet/__init__.py
    full_trace, inner added to both __all__ lists
src/tenet/network/common.py
    scalar, inner DELETED (no re-export, no alias); module keeps spectrum and ones
src/tenet/network/{mps,env,dmrg}.py, src/tenet/network/__init__.py
    callers re-pointed to tenet.full_trace / tenet.inner; __all__ shrinks by two
examples/ctmrg.py, tests/integration/test_{ctmrg,dmrg}.py
    call sites re-pointed
docs/design.md, docs/api/*
    the closed-vs-open principle above; the common.py inventory line updated
```

`ops/contraction.py` is the home, not `ops/basic.py`: `basic.py` opens with "everything here is linear in the blocks, so no categorical machinery is needed", and `inner` needs `adjoint` and `einsum`. `contraction.py` already imports `ops.map`, and putting `full_trace` immediately below `trace` is where a reader looking for one will find the other.

`check_square` moves down to `map_view.py` rather than being imported privately across `ops` modules: its predicate is `as_map(m).codomain.matches(domain)`, pure map-view metadata, and `map_view` is where `as_map` already lives. One move, no copy, no cross-module private reach.

**Every caller, enumerated**

| call site | today | after |
|---|---|---|
| `network/mps.py`:175 `MPS.norm` | `float(scalar(t)) ** 0.5` | `float(tenet.full_trace(t)) ** 0.5` |
| `network/mps.py`:327 `_braket` | `return scalar(t)` | `return tenet.full_trace(t)` |
| `network/env.py`:122 `Env.measure` | `float(scalar(einsum(...)))` | `float(tenet.full_trace(...))` |
| `network/dmrg.py`:65 `lanczos` | `float(inner(vecs[j], w))` | `float(tenet.inner(vecs[j], w))` |
| `examples/ctmrg.py`:180-182 `log_kappa` | three `scalar(...)` | three `tenet.full_trace(...)` |
| `examples/ctmrg.py`:281, :283 `energy` | two `scalar(...)` | two `tenet.full_trace(...)` |
| `tests/integration/test_ctmrg.py`:328 | `tenet.network.scalar(...)` | `tenet.full_trace(...)` |
| `tests/integration/test_dmrg.py`:195 | `network.inner(v, v)` | `tenet.inner(v, v)` |

`expectation_1site` and `expectation_2site` reach `_braket` and need no edit of their own — the two callers #122 counted are covered by one line.

**The hygiene exception: what actually changes, and it is less than hoped**

`tests/network/test_hygiene.py`'s allowed metadata set is `{"qdim", "unit", "provider"}` and it **does not shrink**, because two of the three reads survive: `common.py`:68 `spectrum` still needs `sqrt(qdim)` for its Schmidt weights, and `ctmrg.py`:262 still needs `provider.unit`. What changes is the *justification*, and that must be rewritten in both the test's module docstring and `docs/design.md`:2818-2819, because it becomes false as written:

> ``scalar()`` and ``spectrum()`` need the ``qdim`` weight and **there is no public spelling of a qdim-weighted trace**.

After this PR there is one, `scalar` is gone, and the sole remaining reason for the `qdim` allowance is `spectrum`'s `sqrt(qdim)` diagonal weight. Promoting `spectrum` as well is *not* proposed — it is a Schmidt/corner-spectrum reader with a `sorted(…, reverse=True)` Python list return, which is driver output, not a tensor operation.

Net: `src/tenet/network/` drops one of its three provider reads, and the exception paragraph shortens to one sentence with one owner.

**Acceptance criteria**

- `tenet.full_trace` and `tenet.inner` are in `tenet.__all__` and `tenet.ops.__all__`, reachable as `tenet.full_trace(t)`.
- `tenet.network.scalar` and `tenet.network.inner` no longer exist — not deprecated, not aliased, not re-exported from `network/mps.py`. `network/__init__.__all__` and `network/mps.__all__` shrink accordingly, and the `"from tenet.network.mps import scalar is unchanged for every caller"` line in `mps.py`'s docstring goes with them.
- **Bit-identity, asserted with `==` and not a tolerance**, in the shape #120 and #122 both used: the DMRG sweep energies of `tests/integration/test_dmrg.py`, and `examples/ctmrg.py`'s `log_kappa`, `free_energy`, `energy` and the SGD gradient blocks, are exactly equal before and after. Every current caller passes a square map, so every current caller is arithmetic-identical.
- `full_trace(t) == np.trace(t.to_dense())` for a rank-2 SU(2) map with several qdims in play, and the unweighted sum differs by more than 10% — the same two-part assertion `test_su2_norm_identity_and_the_necessity_of_the_qdim_weight` makes for `norm`.
- `full_trace(t) == np.einsum("abab->", t.to_dense())` for a rank-4 map, pinning that the pairing is the map view and not an axis-adjacent one.
- `full_trace` on `(V OUT, W IN)` with `V != W`, and on a same-side `(V OUT, V OUT)` pair, raises the `_check_square` `ValueError` naming the offending position — the two cases that silently return a number today, each with its current wrong value recorded in the test's comment.
- `full_trace` on a provider without `QuantumDimension` raises `CapabilityError`; on a block-free tensor returns `0.0`.
- `full_trace` under `jax.jit` and `jax.grad` returns a traced scalar and never a Python float; `norm`'s existing JAX test is the template. `inner(a, a) == norm(a)**2` at `rel=1e-12`, moved out of `tests/integration/test_dmrg.py` into `tests/ops/`.
- `tests/network/test_hygiene.py` still passes unchanged in its assertions, with the docstring paragraph rewritten as described; `src/tenet/network/` contains exactly two `provider.qdim`/`provider.unit` reads afterwards.
- No new autoray registration. `ar.do("trace", t)` keeps meaning the partial trace; there is no NumPy name for a qdim-weighted closure to conform to, and `array/dispatch.py`'s table is unchanged.
- The 2956 tests stay green, `tests/network/`'s 10 s and `tests/network/test_ctmrg.py`'s 15 s budgets stand (this PR adds no arithmetic to any loop), and `docs/design.md`'s "Scalars are spelled with boundary legs" section carries the closed-vs-open principle.
- The three-reference survey (quimb, tenpy, YASTN, verified first-hand in this repo's environment as #120 required, plus TensorKit's `tr` cited from source) is in the PR body, with the naming verdict recorded whether or not it agrees with the references — per #120, difference from them is not a bar.

**Out of scope**

- **`spectrum` and `ones` stay in `network/common.py`.** `spectrum` returns a sorted Python list of floats — driver output; `ones` is a seed spelling. Neither is a tensor operation, and `common.py` keeps earning its existence with two shared callers each.
- **A `SymmetricTensor.trace()` method.** `.norm()` exists, so the symmetry argument is available and declined: no caller wants it, and a method is added the day one does.
- **`examples/vmc_mps.py::scalar` is untouched**, as #114's hygiene note and #120 both required — it is `t.blocks[0][0, 0]`, the trivial-leg shortcut, not qdim-weighted, and it must not be conflated. Recorded as a follow-up rather than done here: its `(s, z)` tensor *is* a square map, so `tenet.full_trace` would now return the identical number and the local helper becomes deletable. That is a second, independently-verifiable claim on a differentiated example, and this PR's whole claim is that a move changes nothing.
- **The rank-0 `SymmetricTensor`.** Still open, still requiring an explicit provider field (`structure.py`:71-72's `ponytail:` note). This issue answers the *user's* need for a scalar exit and deliberately does not answer that one.
- **`overlap`, `vdot` or any `inner` variant taking a conjugation flag.** `inner` is sesquilinear in the first argument, full stop.

**Context**

- **In-repo, the code this changes.** `src/tenet/network/common.py`:37-58 (`scalar`, `inner`, and the docstring paragraph that defers this issue by name); `src/tenet/ops/basic.py`:145-171 `norm` (the precedent, its qdim paragraph and its no-`float()` comment, both of which the new docstrings inherit); `src/tenet/ops/contraction.py`:325-347 `trace` and :269-276 (the "no free leg" refusal and its recorded `tenet.norm` precedent — the citation that kills the `axes=None` overload); `src/tenet/ops/linalg.py`:257-285 `_check_square` (moving) and its four callers; `src/tenet/map_view.py`:162-179 `to_matrices` (the body's engine, and why rank-generality is free); `tests/network/test_hygiene.py`:15-32 (the exception paragraph and the `examples/vmc_mps.py` correction); `docs/design.md`:2140-2157 ("Scalars are spelled with boundary legs") and :2808-2819 (the `common.py` inventory and the qdim exception).
- **Issues.** **#114** (moved these four functions with bodies unchanged and deferred this step by name in `common.py`'s docstring), **#120** (the naming bar, the no-aliases rule, the `scalar` keep-and-flag verdict, and the `scalar` → `qdim_trace` parking this issue reverses with an argument), **#122** (the fourth and fifth callers, and the three reasons it kept deferring — public semantics, the open design question, and not wanting three reviews in one PR; all three are answered above), **#112** (the hygiene rules `tenet.network` is held to), **#64** (the keyword-changes-return-type precedent behind the overload refusal), **#95** (backend-genericity: `ar.do` only, torch blocks tested).
- **References for the survey.** TensorKit `tr(t)` for the categorical trace of an endomorphism; quimb, tenpy and YASTN's spellings for full-versus-partial trace and for the inner product, to be verified first-hand in the PR rather than quoted from memory.

Blocks nothing; blocked by nothing open. The `examples/vmc_mps.py` helper deletion and the rank-0 `SymmetricTensor` question are recorded here as follow-ups, not opened.
